core/controlplane/config/encrypted_assets.go

```diff
@@ -378,31 +406,67 @@ func (r *RawAssetsOnMemory) WriteToDir(dirname string, includeCAKey bool) error
 		{"etcd-key.pem", r.EtcdKey, true},
 		{"etcd-client.pem", r.EtcdClientCert, true},
 		{"etcd-client-key.pem", r.EtcdClientKey, true},
+		{"etcd-trusted-ca.pem", r.EtcdTrustedCA, true},
 		{"kubelet-tls-bootstrap-token", r.TLSBootstrapToken, true},
 
 		// Content entirely provided by user, so do not overwrite it if
 		// the file already exists
 		{"tokens.csv", r.AuthTokens, false},
 	}
+
+	if includeCAKey {
+		assets = append(assets, struct {
+			name      string
+			data      []byte
+			overwrite bool
+		}{"ca-key.pem", r.CACert, true})
+	}
+
 	for _, asset := range assets {
 		path := filepath.Join(dirname, asset.name)
 
-		if asset.name != "ca-key.pem" || includeCAKey {
-			if !asset.overwrite {
-				info, err := os.Stat(path)
-				if info != nil {
-					continue
-				}
-
-				// Unexpected error
-				if err != nil && !os.IsNotExist(err) {
-					return err
-				}
+		if !asset.overwrite {
+			info, err := os.Stat(path)
+			if info != nil {
+				continue
 			}
-			if err := ioutil.WriteFile(path, asset.data, 0600); err != nil {
+
+			// Unexpected error
+			if err != nil && !os.IsNotExist(err) {
 				return err
 			}
 		}
+		if err := ioutil.WriteFile(path, asset.data, 0600); err != nil {
+			return err
+		}
+	}
+
+	// etcd trusted ca and worker-ca are separate files, but pointing to ca.pem by default.
+	// In advanced configurations, when certs are managed outside of kube-aws,
+	// these can be separate CAs to ensure that worker nodes have no certs which would let them
+	// access etcd directly. If worker-ca.pem != ca.pem, then ca.pem should include worker-ca.pem
+	// to let TLS bootstrapped workers acces APIServer.
+	symlinks := []struct {
+		from string
+		to   string
+	}{
+		{"ca.pem", "worker-ca.pem"},
+		{"ca.pem", "etcd-trusted-ca.pem"},
+		{"ca-key.pem", "worker-ca-key.pem"},
+	}
+
+	for _, sl := range symlinks {
+		to := filepath.Join(dirname, sl.to)
+
+		if _, err := os.Lstat(to); err == nil {
+			if err := os.Remove(to); err != nil {
+				return err
+			}
+		}
+
+		if err := os.Symlink(sl.from, to); err != nil {
```

Comment:
`to` should be just `sl.to` while calling chdir before/after creating the symlink. Otherwise we get mysterious error when tried to read it:

```
# The file is there
$ ls assets/k8s88/credentials/worker-ca-key.pem
assets/k8s88/credentials/worker-ca-key.pem

# However, actually trying to read it results in a no-such-file error
$ cat assets/k8s88/credentials/worker-ca-key.pem
cat: assets/k8s88/credentials/worker-ca-key.pem: No such file or directory
```

```
$ ls -lah
total 136
drwxr-xr-x  20 mumoshu  staff   680B  9  1 15:29 .
drwxr-xr-x   7 mumoshu  staff   238B  9  1 15:29 ..
-rw-r--r--   1 mumoshu  staff     1B  9  1 15:29 .gitignore
-rw-------   1 mumoshu  staff   1.6K  9  1 15:29 admin-key.pem
-rw-------   1 mumoshu  staff   1.1K  9  1 15:29 admin.pem
-rw-------   1 mumoshu  staff   1.6K  9  1 15:29 apiserver-key.pem
-rw-------   1 mumoshu  staff   1.2K  9  1 15:29 apiserver.pem
-rw-------   1 mumoshu  staff   1.6K  9  1 15:29 ca-key.pem
-rw-------   1 mumoshu  staff   1.0K  9  1 15:29 ca.pem
-rw-------   1 mumoshu  staff   1.6K  9  1 15:29 etcd-client-key.pem
-rw-------   1 mumoshu  staff   1.1K  9  1 15:29 etcd-client.pem
-rw-------   1 mumoshu  staff   1.6K  9  1 15:29 etcd-key.pem
lrwxr-xr-x   1 mumoshu  staff    18B  9  1 15:29 etcd-trusted-ca.pem -> credentials/ca.pem
-rw-------   1 mumoshu  staff   1.1K  9  1 15:29 etcd.pem
-rw-------   1 mumoshu  staff   344B  9  1 15:29 kubelet-tls-bootstrap-token
-rw-------   1 mumoshu  staff     0B  9  1 15:29 tokens.csv
lrwxr-xr-x   1 mumoshu  staff    22B  9  1 15:29 worker-ca-key.pem -> credentials/ca-key.pem
lrwxr-xr-x   1 mumoshu  staff    18B  9  1 15:29 worker-ca.pem -> credentials/ca.pem
-rw-------   1 mumoshu  staff   1.6K  9  1 15:29 worker-key.pem
-rw-------   1 mumoshu  staff   1.1K  9  1 15:29 worker.pem
```

Reply:
Here's my work-around:

```
diff --git a/core/controlplane/config/encrypted_assets.go b/core/controlplane/config/encrypted_assets.go
index b56dc51..3203582 100644
--- a/core/controlplane/config/encrypted_assets.go
+++ b/core/controlplane/config/encrypted_assets.go
@@ -509,8 +509,13 @@ func (r *RawAssetsOnMemory) WriteToDir(dirname string, includeCAKey bool) error
                {"ca-key.pem", "worker-ca-key.pem"},
        }

+       if err := os.Chdir("credentials"); err != nil {
+               return err
+       }
+
        for _, sl := range symlinks {
-               to := filepath.Join(dirname, sl.to)
+               from := sl.from
+               to := sl.to

                if _, err := os.Lstat(to); err == nil {
                        if err := os.Remove(to); err != nil {
@@ -518,10 +523,15 @@ func (r *RawAssetsOnMemory) WriteToDir(dirname string, includeCAKey bool) error
                        }
                }

-               if err := os.Symlink(sl.from, to); err != nil {
+               if err := os.Symlink(from, to); err != nil {
                        return err
                }
        }
+
+       if err := os.Chdir("../"); err != nil {
+               return err
+       }
+
        return nil
 }
```